Combine apiserver and controller-manager into a single service-catalog image

staebler opened the pull request:
Resolves #1239.

* Removes the standalone apiserver and controller-manager binaries.
* Adds a service-catalog binary that can run either the API Server or the Controller Manager.
* Removes the apiserver and controller-manager images.
* Adds a service-catalog image that can run either the API Server or the Controller Manager.
* Updates the helm charts to use the single image for the API Server and the Controller Manager.

To run the API Server, use `service-catalog apiserver`.
To run the Controller Manager, use `service-catalog controller-manager`.

Note that the hyperkube code was copied from k8s/k8s. I could not use the vendored version of the code because it is in a main pacakge.

MHBauer commented on charts/catalog/values.yaml:
```diff
@@ -67,11 +67,6 @@ apiserver:
   serviceAccount: service-catalog-apiserver
 controllerManager:
-  # controller-manager image to use
-  image: quay.io/kubernetes-service-catalog/controller-manager:v0.0.23
```
I think there is still value in being able to specify each image separately. For example: canary apiserver with a specific known version of the controller-manager..

staebler replied:
My issue with this is that it then does not resolve the main crux of the issue, where the user unintentionally uses different versions of apiserver and controller-manager. Who do you see using separate versions and for what purpose? I am imagining that the target audience for that use case would be fairly advanced, perhaps advanced enough to modify the helm charts themselves manually.

Perhaps I could see adding images values for apiserver and controller-manager in addition to the global ones. So if the user sets the values for the specific component, then that would override the global ones. But I do not like the idea of requiring the user to set the value for both components individually in most scenarios.

nilebox replied:
> For example: canary apiserver with a specific known version of the controller-manager..

@MHBauer just make a separate PR with only changes you want, no? I don't see any benefits in testing canary apiserver with release controller-manager, if you changed both in your branch.

pmorie replied:
I'm fine with being able to override it, but I think the chart should use a single image unless you specifically override the api server or controller manager image.

MHBauer commented on cmd/apiserver/app/server/run_server.go:
```diff
@@ -31,15 +31,15 @@ import (
 
 // RunServer runs an API server with configuration according to opts
-func RunServer(opts *ServiceCatalogServerOptions) error {
+func RunServer(opts *ServiceCatalogServerOptions, stopCh <-chan struct{}) error {
```
why an arg vs in options?

staebler replied:
The options are established before the determination is made about which component to run. The channel is not available until the component is run, so we would have to add the channel to the options at that point. Moreover, the channel is not truly an option, and it does not seem appropriate to force it into the options struct.

MHBauer commented on cmd/service-catalog/hyperkube.go:
```diff
@@ -0,0 +1,255 @@
+// servers.
+type HyperKube struct {
+	Name string // The executable name, used for help and soft-link invocation
+	Long string // A long description of the binary.  It will be world wrapped before output.
```
world -> word

staebler replied:
Sure. This was copied from k8s/k8s, but I am happy to fix it here.

MHBauer commented on cmd/service-catalog/hyperkube.go:
```diff
@@ -0,0 +1,255 @@
+		hk.baseFlags.BoolVarP(&hk.helpFlagVal, "help", "h", false, "help for "+hk.Name)
+		hk.baseFlags.BoolVar(&hk.printVersionFlagVal, "version", false, "Print version information and quit")
+		hk.baseFlags.BoolVar(&hk.makeSymlinksFlagVal, "make-symlinks", false, "create a symlink for each server in current directory")
+		hk.baseFlags.MarkHidden("make-symlinks") // hide this flag from appearing in servers' usage output
```
why hidden?

staebler replied:
This was copied from k8s/k8s. If I were to take a guess, I would say because this is really a sub-command rather than a legitimate option.

MHBauer commented on docs/devguide.md:
```diff
@@ -25,8 +25,9 @@ layout:
     │   └── ups-broker          # Helm chart for deploying the user-provided service broker
     ├── cmd                     # Contains "main" Go packages for each service catalog component binary
-    │   └── apiserver           # The service catalog API server binary
-    │   └── controller-manager  # The service catalog controller manager binary
```
If we don't have these directories capable of making binaries, this should probably be moved either to `pkg` or under the `service-catalog` package.

MHBauer commented on docs/devguide.md:
```diff
@@ -277,8 +278,7 @@ helm install ../charts/catalog \
     --set apiserver.tls.cert=$(base64 --wrap 0 ${SC_SERVING_CERT}) \
     --set apiserver.tls.key=$(base64 --wrap 0 ${SC_SERVING_KEY}) \
-    --set apiserver.image=quay.io/kubernetes-service-catalog/apiserver:canary \
-    --set controllerManager.image=quay.io/kubernetes-service-catalog/controller-manager:canary
```
I think we still might want separate values to be able to be set here.

staebler replied:
See above where I argue against this.

duglin commented on charts/catalog/values.yaml:
```diff
@@ -1,12 +1,12 @@
+image: quay.io/kubernetes-service-catalog/service-catalog:v0.0.23
+# imagePullPolicy for the service-catalog; valid values are "IfNotPresent",
+# "Never", and "Always"
+imagePullPolicy: Always
```
I wish we wouldn't default to this too.

staebler replied:
I would think that most of our users in production would want to always pull the latest. What are you suggesting as a better value? IfNotPresent?

duglin replied:
yes I prefer "IfNotPresent". I'm continually running into cases where I already have the image in question locally downloaded and then this flag forces it to be re-fetched, and in some cases pull down a version I didn't want, usually because I did a custom build before hand.  Plus, when I'm on a slow internet (like on a plane), refetching is a killer sometimes.

staebler replied:
My concern is that this makes it more convenient as a developer of the product but more problematic as a user of the product. If that is the case, then I would prefer to keep it as "Always". As a developer, it shouldn't be too bothersome to simply override the imagePullPolicy when running the chart.

staebler replied:
In any event, someone with a dog in this fight should create an issue for this so that we can have a discussion about it outside of this PR.

nilebox commented:
`verify` task fails:
```
Found 1 lint suggestions; failing.
cmd/apiserver/app/server/options.go:90:1: exported method ServiceCatalogServerOptions.AddFlags should have comment or be unexported
```

nilebox commented on cmd/service-catalog/hyperkube.go:
```diff
@@ -0,0 +1,254 @@
+
+// HyperKube represents a single binary that can morph/manage into multiple
+// servers.
+type HyperKube struct {
```
@staebler can you add a comment that this file was copied from the main repo with the path to it?
We should work with apimachinery and core guys on moving all shared code to separate repo.

pmorie replied:
I'm afraid I cannot accept any other name than `HyperKat`

nilebox commented on cmd/service-catalog/template.go:
```diff
@@ -14,6 +14,10 @@ See the License for the specific language governing permissions and
 
+// This file is copied from k8s.io/kubernetes/pkg/util/template. The file
+// needs to be copied because it is from the 1.8 version of kubernetes. Once
+// the move to kubernetes 1.8 is completed, this file can be removed.
```
re: after 1.8
depends whether we want to keep dependency on kubernetes main repo, see https://github.com/kubernetes-incubator/service-catalog/issues/1323

nilebox commented:
One concern with running both API server and controller manager in a single Docker image is that then the logging will be joined together. It will be harder to analyze logs.
Not sure if there are any techniques mitigating this.

Also the HyperKube logic making the code more complicated concerns me too.

nilebox commented:
Ah I think I got it wrong, so API server and controller manager will remain running as separate Pods, just both will be part of the same Docker image?

nilebox commented:
Ok, reading https://stackoverflow.com/questions/33953254/what-is-hyperkube was helpful

staebler commented:
> Ah I think I got it wrong, so API server and controller manager will remain running as separate Pods, just both will be part of the same Docker image?

Correct. All this is doing is combining the API server and the controller manager into a single image. It does not change anything about how anything is run other than the mechanism by which they are started. The purpose of this is to reduce the risk of the end user accidentally running different versions of API server and controller manager, where we do not make any effort to make different versions compatible with each other.

As for the complexity of the hyperkube, I don't think that it makes the service catalog code any more complex. If anything, there was complexity removed from that code. We are inherited complexity from the hyperkube code itself, but that can hopefully be removed in the future by vendoring in the code from k8s instead of copying it into the service catalog tree.

nilebox commented:
> We are inherited complexity from the hyperkube code itself, but that can hopefully be removed in the future by vendoring in the code from k8s instead of copying it into the service catalog tree.

I am still not sure if depending on k8s/k8s is a good idea. To make it easier to keep track of all the changes happening to the types we copied from it, I would probably just created a special package `service-catalog/pkg/kubernetes/*` with the structure replicating the k8s/k8s one. Then for each update on the new k8s release I would overwrite those files with the latest version.

WDYT?

staebler commented:
> I am still not sure if depending on k8s/k8s is a good idea.

I agree that it is too much to add a dependency on k8s/k8s just for the hyperkube. What I meant was that hopefully we can work to pull the hyperkube functionality out into a smaller dependency that would be appropriate for service catalog to consume.

staebler commented:
I have moved the code that was copied from k8s.io/kubernetes and k8s.io/apiserver to pkg/kubernetes and pkg/kube-apiserver, respectively. The file structure under those packages mimics the file structure in the original trees. I have added README.md files in the roots of those two new packages to document where the files came from and what changes have been made to the files.

nilebox commented:
LGTM
> I agree that it is too much to add a dependency on k8s/k8s just for the hyperkube

Actually we already have this dependency and I would like to remove it using the same technique as you did for HyperKube, see https://github.com/kubernetes-incubator/service-catalog/issues/1323

pmorie commented:
sorry, @staebler, needs rebase

pmorie commented on cmd/apiserver/app/server/options.go:
```diff
@@ -26,6 +26,22 @@ import (
+	// k8s core API server.
+	certDirectory = "/var/run/kubernetes-service-catalog"
+
+	// I made this up to match some existing paths. I am not sure if there
```
etcdPathPrefix and GroupName need to come back out - see #1394

pmorie commented on pkg/hyperkube/README.md:
```diff
@@ -0,0 +1,15 @@
+This package contains files copied from the main [kubernetes/kubernetes repo](https://github.com/kubernetes/kubernetes).
```
Is there an upstream issue yet to move these out to another repo?

staebler replied:
https://github.com/kubernetes/kubernetes/issues/54299

pmorie commented:
💥
